src/Utilities/Compiler/Extensions/IOperationExtensions.cs

```diff
@@ -1136,6 +1136,20 @@ public static bool IsAnyCompoundAssignment(this IOperation operation)
                 _ => false,
             };
 #endif
+
+        public static ISymbol? GetSymbolFromReference(this IOperation reference)
+        {
+            return reference switch
+            {
+                ILocalReferenceOperation l => l.Local,
+                IFieldReferenceOperation f => f.Field,
+                IParameterReferenceOperation pa => pa.Parameter,
+                IPropertyReferenceOperation p => p.Property,
+                IMemberReferenceOperation m => m.Member,
+                IArrayElementReferenceOperation a => GetSymbolFromReference(a.ArrayReference),
```

Comment:
this would be true for data[0] and data[1], since it only compares that data is the same symbol.
I refactored my code to handle the indices directly
https://github.com/dotnet/roslyn-analyzers/blob/41da30d77f263148615211dd61945c94a0da5487/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryTryGetValueAnalyzer.cs#L384-L394

Reply:
Can you explain why this would only be true for `data[0]` and `data[1]`? From what I can tell, it fetches the array reference symbol regardless of index, which is exactly the desired behavior.

Reply:
``` c#
var data = new int[]{1,2};
var x = data[0];
var y = data[1];
```
If you look at `data[0]`, you will get `IArrayElementReferenceOperation` with `ArrayReference` being a `ILocalReferenceOperation` where the `Local` symbol is `data`, but the symbol itself does not tell you how it was used, data[1] has the same Symbol in its `IArrayElementReferenceOperation.ArrayReference.Local`.

To make sure that the operation not only reference the same array but also the same element, you need to test `IArrayElementReferenceOperation.Indices`.
Which here would be `ILiteralOperation` with ConstantValue 0 and 1 respectively, but could also be another reference operation like `data[data[0]]`.